File: lab2/parser.c
#include<stdio.h>
#include<string.h>
#include<stdlib.h>

#define RED "\x1B[31m"
#define RESET "\x1B[0m"

typedef struct {
	char tip[16];
	int redni_broj;
} Zapis;

int predvidiTip(int paritet, int redni_broj){
	if(redni_broj % 2 != 0) {
		if(paritet == 0) return 1;
		else return 0;
	}
	else return paritet;
}

int main(int argc, char* argv[]){
	
	if(argc < 2){
		printf(RED "Neispravan poziv naredbe!\n" RESET);
		printf("%s <ulazna_datoteka> <debug?>\n", argv[0]);
		exit(-1);
	}
	
	int debug = argc == 3 ? strcmp(argv[2], "debug") == 0 ? 1 : 0 : 0;
	
	char putanja[64] = {0};
	
	if(sscanf(argv[1], "%s", putanja) != 1){
		printf("Neispravan unos putanje datoteke!\n");
		exit(-1);
	}
	
	FILE* datoteka = fopen(putanja, "r");
	
	if(datoteka == NULL){
		printf("Nije moguce otvoriti datoteku '%s'\n", putanja);
		exit(-2);
	}
	
	printf("Citanje datoteke '%s' ...\n", putanja);
	
	char tip[2][16] = {"Roditelj", "Dijete"};
	int redni_broj = 0, broj_neocekivanih = 0, paritet = 0;
	Zapis zapis, prethodni;
	
	fscanf(datoteka, "%s #%d", zapis.tip, &zapis.redni_broj);
	
	if(strcmp(zapis.tip, tip[0]) == 0) paritet = 0;
	else if (strcmp(zapis.tip, tip[1]) == 0) paritet = 1;
	else {
		printf("Nesto nije u redu s ulaznom datotekom! Napraviti rucnu provjeru...\n");
		exit(-3);
	}
	
	prethodni = zapis;
	redni_broj++;
	
	while(fscanf(datoteka, "%s #%d", zapis.tip, &zapis.redni_broj) == 2){
		redni_broj++;

		if(strcmp(zapis.tip, tip[predvidiTip(paritet, redni_broj)]) == 0){
			broj_neocekivanih++;
			if(debug) printf("Neocekivani zapis (%d): %s #%d\n", redni_broj, zapis.tip, zapis.redni_broj);
		}

		prethodni = zapis;
	}
	
	if(redni_broj % 2 == 0 && broj_neocekivanih % 2 == 0){
		broj_neocekivanih /= 2;
		redni_broj /= 2;
	}
	else {
		printf("Nesto nije u redu s ulaznom datotekom! (%d/%d) Napraviti rucnu provjeru...\n", broj_neocekivanih, redni_broj);
		exit(-3);
	}
	
	printf("Broj odstupanja: %d\n", broj_neocekivanih);
	printf("Broj izvodjenja: %d\n", redni_broj);
	
	if(broj_neocekivanih > 0){
		double odstupanje = (double)broj_neocekivanih * 100 / redni_broj;
		printf(RED "\nOdstupanje: %.3lf\%\n" RESET, odstupanje);
		printf("Točnost: %.3lf\%\n", 100.00 - odstupanje);
	}
	
	fclose(datoteka);
	
	return 0;
}

File: lab2/parser.py
import subprocess, sys, os

class Zapis:
	def __init__(self, tip:str, redni_broj:int):
		self.tip = tip
		self.redni_broj = redni_broj
		
class Mjerenje:
	def __init__(self):
		self.velicina = 0
		self.odstupanja = 0
		self.paritet = 0
		self.tipovi = ["Roditelj", "Dijete"]
	
	def odstupanje(self):
		return float(self.odstupanja) * 100 / self.velicina
	
	def tocnost(self):
		return 100 - self.odstupanje()
		
	def postavi_paritet(self, zapis:Zapis):
		if zapis.tip == self.tipovi[0]:
			self.paritet = 0
		elif zapis.tip == self.tipovi[1]:
			self.paritet = 1
		else:
			raise ValueError("Nesto nije u redu s ulaznom datotekom! Napravite rucnu provjeru...")
		
	def provjeri_zapis(self, zapis:Zapis):
		self.velicina += 1
		
		if self.velicina == 1:
			self.postavi_paritet(zapis)
		elif zapis.tip != self.predvidi_tip_zapisa():
			self.odstupanja += 1
		
	def zavrsi(self):
		if self.velicina % 2 == 0 and self.odstupanja % 2 == 0:
			self.odstupanja //= 2
			self.velicina //=  2
		else:
			raise ValueError("Nesto nije u redu s ulaznom datotekom! Napravite rucnu provjeru...")

	def predvidi_tip_zapisa(self):
		if self.velicina % 2 == 0:
			if self.paritet == 0:
				return self.tipovi[1]
			else:
				return self.tipovi[0]
		else:
			return self.tipovi[self.paritet]

if len(sys.argv) >= 2 and sys.argv[1].isnumeric():
	ponavljanje = int(sys.argv[1])
else:
	ponavljanje = 1
	
if os.path.exists("output") == False:
	os.mkdir("output")
	
mjerenja = []
	
for i in range(0,ponavljanje):
	
	print("\n{}. iteracija:".format(i+1))
	
	for velicina in [10, 100, 1000, 10000, 100000]:
		try:
			putanja = "output/mjerenje-{}.file".format(velicina)
			datoteka = open(putanja, "w+")
			
			print("\n{:=^30}".format(" Mjerenje za {} djece ".format(velicina)))
			
			subprocess.run(["./lab2.out", str(velicina)], stdout=datoteka)
			datoteka.seek(0)

			mjerenje = Mjerenje()

			for linija in datoteka:
				tip, broj = linija.strip().split(" #")
				mjerenje.provjeri_zapis(Zapis(tip, int(broj)))

			mjerenje.zavrsi()
			
			print("Broj odstupanja: {odstupanja:d}".format(odstupanja=mjerenje.odstupanja))
			print("Broj izvodjenja: {izvodjenja:d}".format(izvodjenja=mjerenje.velicina))

			if mjerenje.odstupanja > 0:
				print("\nOdstupanje: {odstupanje:.3f}%".format(odstupanje=mjerenje.odstupanje()))
				print("Tocnost: {tocnost:.3f}%".format(tocnost=mjerenje.tocnost()))

			print("{:=^30}".format(" Kraj mjerenja ".format(velicina)))
			datoteka.close()
			mjerenja.append(mjerenje)

		except FileNotFoundError:
			print("Nije moguće otvoriti datoteku!")
		except ValueError as ex:
			print(ex.message)

mjerenja.sort(key=lambda mjerenje: mjerenje.velicina)

print("\n[{:*^28}]".format(" Sažetak mjerenja "))

for velicina in [10, 100, 1000, 10000, 100000]:
	print("{:=^30}".format(" {} procesa djece ".format(velicina)))
	print("{:=^20}{:=^10}".format(" odstupanja ", " točnost "))
	print("{:^10}{:^10}{:^10}".format(" broj ", " % ", " % "))
	
	for mjerenje in filter(lambda m: m.velicina == velicina, mjerenja):
		print("{:^10}{:^10.3f}{:^10.3f}".format(mjerenje.odstupanja, mjerenje.odstupanje(), mjerenje.tocnost()))
	print("="*30)
